### ISSUE TITLE
ROCm build lacks ridiculous amount of dependencies

### ISSUE THREAD

**Issue by Column01** [pytorch/pytorch#63188]
## 📚 Documentation

When following the build from source guide to build the ROCm version of pytorch (see https://github.com/pytorch/pytorch/issues/53738 for why i'm even doing that in the first place), the build will fail for multiple missing dependencies not listed in the guide for building for AMD. There are also a few environment variables that are missing for the build (see https://github.com/pytorch/pytorch/issues/62536) and I had no luck setting `ROCclr_DIR` directly to the config file the build is asking for. I even built is manually following the [AMD guide](https://rocmdocs.amd.com/en/latest/Installation_Guide/HIP-Installation.html#build-rocclr) (which also sets that variable) and pytorch still whines about it not being present when building.

Short list of those I could find were missing from the docs for building:
- `rocblas`
- `rocrand`
- `hipfft`
- `miopen-hip`
- probably others but I gave up after spending 6 hours trying to build the program.

## Potential Solutions

Update the docs to include a **proper** guide to building on ROCm wheels on linux that includes installing dependencies and setting environment variables for the build process

OR

Update the build process to include building wheels for `gfx803` as was intended (I think anyways, it's mentioned in the `LoadHIP.cmake` as a supported architecture) but a small oversight does not allow (again, see https://github.com/pytorch/pytorch/issues/53738 and the corresponding PR to fix it: https://github.com/pytorch/pytorch/pull/54653)

**Comment by jithunnair-amd** [comment]
Re: gfx803 support, please refer https://github.com/pytorch/pytorch/issues/53738#issuecomment-899970052

**Comment by Column01** [comment]
still need the build guide fixed

**Comment by jithunnair-amd** [comment]
Just to make sure we're using the same source, can you post the link to the guide you're referring to? Thanks!

**Comment by Column01** [comment]
> Just to make sure we're using the same source, can you post the link to the guide you're referring to? Thanks!

https://github.com/pytorch/pytorch#from-source

The guide in general is pretty terrible for someone who's never built code before (not my case). It's written in a way to minimize duplicating the parts of the guide that would be repeated for each, but in the process, it's also extremely hard to follow. The main issue with it though is the install dependencies section, it only covers adding the ones for cuda, but doesn't list the miriad of requirements needed for ROCm.

My recommendation would be to create a separate section for each OS and in each, go through installing the dependencies for CPU, CUDA and ROCm builds (the latter only applies to Linux) and then finally building them.

Regularly during my attempt to build it, it would fail and say something along the lines of cmake not being able to locate a config for x module (most of which are listed above).

**Comment by seemethere** [comment]
> > Just to make sure we're using the same source, can you post the link to the guide you're referring to? Thanks!
> 
> [pytorch/pytorch#from-source](https://github.com/pytorch/pytorch#from-source)
> 
> The guide in general is pretty terrible for someone who's never built code before (not my case). It's written in a way to minimize duplicating the parts of the guide that would be repeated for each, but in the process, it's also extremely hard to follow. The main issue with it though is the install dependencies section, it only covers adding the ones for cuda, but doesn't list the miriad of requirements needed for ROCm.
> 
> My recommendation would be to create a separate section for each OS and in each, go through installing the dependencies for CPU, CUDA and ROCm builds (the latter only applies to Linux) and then finally building them.
> 
> Regularly during my attempt to build it, it would fail and say something along the lines of cmake not being able to locate a config for x module (most of which are listed above).

We'd appreciate any contributions you or others have to make it better / less confusing

Also if this this is better served in a separate file rather than in the main README.md I'm all for that as well

**Comment by Column01** [comment]
Well I'd love to help update the guide, I wasn't even able to build it on my machine so not sure how I'd be able to help.

It would be ideal if the guide were in a separate file though

**Comment by Column01** [comment]
I noticed there is a build script in .jenkins/pytorch/build.sh`, is it worth me trying to run it manually? It appears to just follow the same steps that the build guide does.

**Comment by jeffdaily** [comment]
> I noticed there is a build script in `.jenkins/pytorch/build.sh`, is it worth me trying to run it manually? It appears to just follow the same steps that the build guide does.

That script can be run manually, but it expects to be run inside a docker image that was prepared for CI purposes.  There are a lot of assumptions about your installed packages and environment when using that script.  For the script to work at all you must set the `BUILD_ENVIRONMENT` env var to [one of the names from this case statement](https://github.com/pytorch/pytorch/blob/7701ea48be276f5058cc6247ec53fb6a5789445a/.circleci/docker/build.sh#L75-L251) in the list of CI images.

You can prepare a CI image by running the build.sh script I linked to above and pass it the name of one of the images you want, and tag it something useful to you.  For example

```
cd .circleci/docker
./build.sh pytorch-linux-bionic-rocm4.2-py3.6 -t pytorch-rocm4.2-py3.6-CI-base-image
```

Then you can run the image and the `.jenkins/pytorch/build.sh` script would work as-is in that container.

**Comment by Column01** [comment]
> That script can be run manually, but it expects to be run inside a docker image that was prepared for CI purposes. There are a lot of assumptions about your installed packages and environment when using that script. For the script to work at all you must set the `BUILD_ENVIRONMENT` env var to [one of the names from this case statement](https://github.com/pytorch/pytorch/blob/7701ea48be276f5058cc6247ec53fb6a5789445a/.circleci/docker/build.sh#L75-L251) in the list of CI images.

Those same assumptions appear to have been made with the normal build guide 😁
